File: diffdesk/static/junction.js
// 多対多(親A - 中間 - 親B)の移行検証タブ。
import { $, $$, api, apiJson, downloadResponse, escapeHtml, postJson, toast } from "/static/common.js";

let jxFiles = [];  // /api/files の結果

async function refreshJxFiles() {
  try {
    const { files } = await apiJson("/api/files");
    jxFiles = files;
    $$(".jx-file").forEach(sel => {
      const cur = sel.value;
      sel.innerHTML = `<option value="">(選択)</option>` +
        files.map(f =>
          `<option value="${f.file_id}">${escapeHtml(f.filename)} (${f.total_rows}行)</option>`
        ).join("");
      if (cur && files.some(f => f.file_id === cur)) sel.value = cur;
      fillColumns(sel);
    });
  } catch (e) { toast(e.message, true); }
}

function columnsOf(fileId) {
  return jxFiles.find(f => f.file_id === fileId)?.columns ?? [];
}

const jxDesired = {};  // 移行定義JSONから読み込んだ希望値(列が現れたら自動選択)

function fillSelect(sel, columns, { optional = false } = {}) {
  const cur = sel.value;
  sel.innerHTML = (optional ? `<option value="">(${sel.id === "jx-required" ? "なし" : "確認しない"})</option>` : "") +
    columns.map(c => `<option value="${escapeHtml(c)}">${escapeHtml(c)}</option>`).join("");
  const want = jxDesired[sel.id];
  if (want && columns.includes(want)) sel.value = want;
  else if (cur && columns.includes(cur)) sel.value = cur;
}

function fillColumns(fileSel) {
  const cols = columnsOf(fileSel.value);
  switch (fileSel.id) {
    case "jx-file-source":
      fillSelect($("#jx-a-source"), cols);
      fillSelect($("#jx-b-source"), cols);
      fillSelect($("#jx-required"), cols, { optional: true });
      break;
    case "jx-file-a":
      fillSelect($("#jx-a-ext"), cols);
      break;
    case "jx-file-b":
      fillSelect($("#jx-b-ext"), cols);
      break;
    case "jx-file-j":
      fillSelect($("#jx-j-key"), cols);
      fillSelect($("#jx-ref-a"), cols, { optional: true });
      fillSelect($("#jx-ref-b"), cols, { optional: true });
      break;
  }
}

function jxRequest() {
  if (!$("#jx-file-source").value || !$("#jx-file-j").value) {
    toast("移行元データと中間抽出の2つのファイルを選択してください。", true);
    return null;
  }
  return {
    file_source: $("#jx-file-source").value,
    file_a: $("#jx-file-a").value,
    file_b: $("#jx-file-b").value,
    file_j: $("#jx-file-j").value,
    config: {
      a_source_col: $("#jx-a-source").value,
      b_source_col: $("#jx-b-source").value,
      a_ext_col: $("#jx-a-ext").value,
      b_ext_col: $("#jx-b-ext").value,
      j_key_col: $("#jx-j-key").value,
      key_template: $("#jx-template").value.trim() || "{A}-{B}",
      a_regex_pattern: $("#jx-a-pattern").value,
      a_regex_replacement: $("#jx-a-repl").value,
      b_regex_pattern: $("#jx-b-pattern").value,
      b_regex_replacement: $("#jx-b-repl").value,
      required_col: $("#jx-required").value,
      j_ref_a_col: $("#jx-ref-a").value,
      j_ref_b_col: $("#jx-ref-b").value,
    },
  };
}

const CAUSE_JA = {
  missing_a: "親Aの外部IDが存在しない",
  missing_b: "親Bの外部IDが存在しない",
  missing_both: "親A・親Bの両方が存在しない",
  parents_ok: "両親は存在(中間だけ未取込)",
  unknown: "原因不明(詳細設定で親の抽出ファイルを選ぶと切り分けできます)",
};
const CAUSE_CLS = { missing_a: "jx-c-a", missing_b: "jx-c-b",
                    missing_both: "jx-c-both", parents_ok: "jx-c-j",
                    unknown: "jx-c-u" };

// ---------------------------------------------------------------- 関係ビュー
let relState = { groups: [], page: 0, perPage: 50 };

function relFiltered() {
  const onlyNg = $("#jx-rel-problems").checked;
  const q = $("#jx-rel-search").value.trim().toLowerCase();
  return relState.groups.filter(g => {
    if (onlyNg && g.ng === 0) return false;
    if (!q) return true;
    if (g.a.toLowerCase().includes(q)) return true;
    return g.items.some(i => i.b.toLowerCase().includes(q)
                          || i.key.toLowerCase().includes(q));
  });
}

function renderRelations() {
  const list = relFiltered();
  const per = relState.perPage;
  const pages = Math.max(1, Math.ceil(list.length / per));
  relState.page = Math.min(relState.page, pages - 1);
  const slice = list.slice(relState.page * per, (relState.page + 1) * per);
  $("#jx-rel-page-info").textContent =
    list.length ? `${relState.page * per + 1}–${relState.page * per + slice.length} / ${list.length}親` : "0件";
  if (!slice.length) {
    $("#jx-relations").innerHTML =
      `<p class="hint">${relState.groups.length
        ? "条件に合う親がありません(「問題のある親だけ表示」を外すと全件出ます)。"
        : "関係がありません。"}</p>`;
    return;
  }
  $("#jx-relations").innerHTML = slice.map(g => {
    const chips = g.items.map(i => {
      if (i.status === "ok") {
        return `<span class="relchip rel-ok" title="取込済: ${escapeHtml(i.key)}">✔ ${escapeHtml(i.b)}</span>`;
      }
      if (i.status === "extra") {
        return `<span class="relchip rel-extra" title="移行元に無い実績: ${escapeHtml(i.key)}">? ${escapeHtml(i.b)}</span>`;
      }
      return `<span class="relchip rel-missing" title="未取込: ${escapeHtml(i.key)} — ${CAUSE_JA[i.cause] || i.cause}">✖ ${escapeHtml(i.b)}</span>`;
    }).join(" ");
    const stat = g.ng
      ? `<span class="ng-mark">✖ ${g.ok}/${g.items.length} 取込済</span>`
      : `<span class="ok-mark">✔ 全${g.items.length}件 取込済</span>`;
    return `<div class="relgroup"><div class="relhead"><b>${escapeHtml(g.a)}</b> ${stat}</div>` +
      `<div class="relbody">${chips}</div></div>`;
  }).join("");
}

$("#jx-rel-problems").onchange = () => { relState.page = 0; renderRelations(); };
$("#jx-rel-search").oninput = () => { relState.page = 0; renderRelations(); };
$("#jx-rel-prev").onclick = () => { relState.page = Math.max(0, relState.page - 1); renderRelations(); };
$("#jx-rel-next").onclick = () => { relState.page += 1; renderRelations(); };

function macroRow(step, extraNote = "") {
  const mark = step.passed
    ? `<span class="ok-mark">✔</span>`
    : `<span class="ng-mark">✖</span>`;
  const notes = [];
  if (step.missing_count) {
    notes.push(`欠落${step.missing_count}件(例: ${step.missing.slice(0, 5).map(escapeHtml).join(", ")})`);
  }
  if (step.extra_count) {
    notes.push(`想定外${step.extra_count}件(例: ${step.extra.slice(0, 5).map(escapeHtml).join(", ")})`);
  }
  if (step.dup_actual) notes.push(`抽出側キー重複${step.dup_actual}件`);
  if (extraNote) notes.push(extraNote);
  return `<tr><td>${mark}</td><td><b>${escapeHtml(step.name)}</b></td>` +
    `<td>${step.expected}</td><td>${step.actual}</td>` +
    `<td>${step.diff > 0 ? "+" : ""}${step.diff}</td>` +
    `<td class="hint">${notes.join(" / ") || "—"}</td></tr>`;
}

function renderJxResult(r) {
  $("#jx-result").hidden = false;
  const banner = $("#jx-banner");
  banner.className = r.passed ? "ok" : "ng";
  banner.textContent = r.passed
    ? "✔ 移行OK — 期待どおり取り込まれています"
    : "✖ 要確認 — 期待値と実績に差があります(関係ビューと下の内訳を確認)";

  // 関係ビュー
  relState = { ...relState, groups: r.relations?.groups ?? [], page: 0 };
  const rel = r.relations;
  renderRelations();
  if (rel && (rel.truncated || rel.unparsed_extra_count)) {
    const notes = [];
    if (rel.truncated) notes.push(`親が多いため先頭${rel.groups.length}件のみ表示(全${rel.total_groups}件)`);
    if (rel.unparsed_extra_count) {
      notes.push(`複合キー形式に合わない実績キー${rel.unparsed_extra_count}件: ` +
        rel.unparsed_extra.slice(0, 5).map(escapeHtml).join(", "));
    }
    $("#jx-relations").insertAdjacentHTML("beforebegin",
      `<p class="hint jx-rel-note">${notes.join(" / ")}</p>`);
  }
  $$(".jx-rel-note").forEach((el, i, all) => { if (i < all.length - 1) el.remove(); });

  const j = r.junction;
  const skipNote =
    `スキップ: 必須列空白${j.skipped_required}件・キー成分空白${j.skipped_empty_pair}件`;
  const skippedParents = [];
  if (!r.parent_a) skippedParents.push("親A");
  if (!r.parent_b) skippedParents.push("親B");
  $("#jx-macro").innerHTML =
    `<table><thead><tr><th>判定</th><th>ステップ</th><th>期待値(ユニーク)</th>` +
    `<th>実績件数</th><th>差</th><th>内訳</th></tr></thead><tbody>` +
    (r.parent_a ? macroRow(r.parent_a) : "") +
    (r.parent_b ? macroRow(r.parent_b) : "") +
    macroRow(j, skipNote) +
    `</tbody></table>` +
    (skippedParents.length
      ? `<p class="hint">${skippedParents.join("・")}の件数検証はスキップ` +
        `(詳細設定で抽出ファイルを選ぶと有効になります)</p>` : "") +
    (r.ref_errors
      ? `<p class="hint">参照整合: 親A参照エラー <b>${r.ref_errors.bad_ref_a}</b>件 / ` +
        `親B参照エラー <b>${r.ref_errors.bad_ref_b}</b>件</p>`
      : "");

  const o = r.orphans;
  $("#jx-orphan-card").hidden = !o.total;
  $("#jx-export").hidden = !o.total;
  if (o.total) {
    const order = ["missing_a", "missing_b", "missing_both", "parents_ok", "unknown"];
    const bar = order.map(c => o.causes[c]
      ? `<div class="${CAUSE_CLS[c]}" style="width:${(o.causes[c] / o.total) * 100}%" ` +
        `title="${CAUSE_JA[c]}: ${o.causes[c]}件"></div>` : "").join("");
    const legend = order.filter(c => o.causes[c]).map(c =>
      `<span><i class="${CAUSE_CLS[c]}"></i>${CAUSE_JA[c]}: <b>${o.causes[c]}</b>件</span>`
    ).join("");
    const bn = o.bottleneck
      ? `<p><b>ボトルネック: ${CAUSE_JA[o.bottleneck]}</b>(未取込${o.total}件のうち最多)</p>`
      : "";
    const sample =
      `<table><thead><tr><th>複合キー</th><th>親Aキー</th><th>親Bキー</th><th>原因</th></tr></thead><tbody>` +
      o.samples.slice(0, 30).map(s =>
        `<tr><td>${escapeHtml(s.key)}</td><td>${escapeHtml(s.a)}</td>` +
        `<td>${escapeHtml(s.b)}</td><td>${CAUSE_JA[s.cause]}</td></tr>`).join("") +
      `</tbody></table>` +
      (o.total > 30 ? `<p class="hint">全${o.total}件は「未取込一覧CSV」で出力できます。</p>` : "");
    $("#jx-orphans").innerHTML =
      bn + `<div class="jx-causebar">${bar}</div>` +
      `<div class="jx-legend">${legend}</div>` + sample;
  }
}

$("#jx-run").onclick = async () => {
  const req = jxRequest();
  if (!req) return;
  try {
    const r = await postJson("/api/junction-verify", req);
    renderJxResult(r);
    toast("多対多検証を実行しました。");
  } catch (e) { toast(e.message, true); }
};

$("#jx-export").onclick = async () => {
  const req = jxRequest();
  if (!req) return;
  try {
    const res = await api("/api/junction-verify/orphans", {
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ ...req, encoding: "utf-8-sig" }),
    });
    await downloadResponse(res, "未取込一覧.csv");
  } catch (e) { toast(e.message, true); }
};

// 移行定義JSON(仕様書)から設定を自動入力
$("#jx-spec-btn").onclick = () => $("#jx-spec-input").click();
$("#jx-spec-input").onchange = async () => {
  const files = [...$("#jx-spec-input").files];
  $("#jx-spec-input").value = "";
  if (!files.length) return;
  try {
    const specs = [];
    for (const f of files) specs.push(JSON.parse(await f.text()));
    const r = await postJson("/api/migration-spec/junction", { specs });
    const s = r.settings;
    $("#jx-template").value = s.key_template;
    $("#jx-a-pattern").value = s.a_regex_pattern || "";
    $("#jx-a-repl").value = s.a_regex_replacement || "";
    $("#jx-b-pattern").value = s.b_regex_pattern;
    $("#jx-b-repl").value = s.b_regex_replacement;
    // セレクトは希望値として保持し、該当列があるファイルを選んだ時点で自動選択
    const map = {
      "jx-a-source": s.a_source_col, "jx-b-source": s.b_source_col,
      "jx-required": s.required_col,
      "jx-a-ext": s.a_ext_col, "jx-b-ext": s.b_ext_col,
      "jx-j-key": s.j_key_col,
      "jx-ref-a": s.j_ref_a_col, "jx-ref-b": s.j_ref_b_col,
    };
    for (const [id, val] of Object.entries(map)) {
      if (!val) continue;
      jxDesired[id] = val;
      const sel = $("#" + id);
      if ([...sel.options].some(o => o.value === val)) sel.value = val;
    }
    $("#jx-spec-info").textContent =
      `定義「${r.object || "中間オブジェクト"}」を読み込みました。` +
      `ファイルを選ぶと該当する列が自動選択されます。`;
    toast("移行定義JSONから設定を読み込みました。" +
      (r.warnings.length ? " 注意: " + r.warnings.join(" / ") : ""),
      r.warnings.length > 0);
  } catch (e) { toast(`定義JSONの読み込みに失敗: ${e.message}`, true); }
};

// 複合キー形式の自動推定(実例から)
$("#jx-infer").onclick = async () => {
  const src = $("#jx-file-source").value;
  const jf = $("#jx-file-j").value;
  if (!src || !jf) return toast("移行元データと中間抽出を先に選択してください。", true);
  if (!$("#jx-a-source").value || !$("#jx-b-source").value || !$("#jx-j-key").value) {
    return toast("親A・親Bの列と中間キーの列を先に選択してください。", true);
  }
  try {
    const r = await postJson("/api/junction-verify/infer-template", {
      file_source: src, file_j: jf,
      a_source_col: $("#jx-a-source").value,
      b_source_col: $("#jx-b-source").value,
      j_key_col: $("#jx-j-key").value,
      a_regex_pattern: $("#jx-a-pattern").value,
      a_regex_replacement: $("#jx-a-repl").value,
      b_regex_pattern: $("#jx-b-pattern").value,
      b_regex_replacement: $("#jx-b-repl").value,
    });
    if (!r.template) {
      $("#jx-infer-info").textContent = "推定できませんでした(値がキーに含まれない形式かもしれません)";
      return toast("キー形式を推定できませんでした。手で入力してください。", true);
    }
    $("#jx-template").value = r.template;
    const pct = Math.round(r.coverage * 100);
    $("#jx-infer-info").textContent = `推定: ${r.template}(サンプル${r.checked}組中${pct}%が実例と一致)`;
    toast(`複合キーの形式を「${r.template}」と推定しました(一致率${pct}%)。`);
  } catch (e) { toast(e.message, true); }
};

$("#jx-refresh").onclick = refreshJxFiles;
$$(".jx-file").forEach(sel => sel.onchange = () => fillColumns(sel));
// タブを開いたときにファイル一覧を最新化
document.querySelector('#tabs button[data-tab="tab-junction"]')
  .addEventListener("click", refreshJxFiles);

File: diffdesk/core/junction.py
"""多対多(親A - 中間 - 親B)モデルの移行検証。

データ移行QAの定型ロジックを実装する:
  ステップ1: 親Aのマクロ検証 — 移行元のユニーク件数(期待値) vs 抽出の実績件数
  ステップ2: 親Bのマクロ検証 — 空白除外+正規化後のユニーク件数 vs 実績件数
  ステップ3: 中間のマクロ検証 — スキップ条件適用後の複合キーのユニーク件数 vs 実績件数
  ステップ4: 未取込の孤立分析 — 欠落した中間キーの原因を
             「親A欠落 / 親B欠落 / 両親欠落 / 親は存在(中間のみ欠落)」に切り分け

すべて純Python。比較は normalize の既定正規化(空白除去・NFKC・数値同一視)を通す。
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field

from .model import DiffDeskError, DiffOptions, Table
from .normalize import make_normalizer

_SAMPLE_LIMIT = 200  # レポートに載せる欠落・過剰キーの上限


@dataclass
class JunctionConfig:
    """多対多検証の設定。列名は各ファイルの実列名。"""

    # 移行元(1ファイルに親A・親Bのソース値が行単位で入っている前提)
    a_source_col: str            # 親Aのソース列
    b_source_col: str            # 親Bのソース列
    # 抽出側
    a_ext_col: str               # 親A抽出の外部ID列
    b_ext_col: str               # 親B抽出の外部ID列
    j_key_col: str               # 中間抽出の複合キー列
    # 複合キー生成テンプレート({A} {B} を置換)
    key_template: str = "{A}-{B}"
    # 親Bの正規化(任意の正規表現置換)
    b_regex_pattern: str = ""
    b_regex_replacement: str = ""
    # スキップ条件: この列が空白の行は取り込み対象外(任意)
    required_col: str = ""
    # 中間抽出の親参照列(任意・指定時は参照整合も確認)
    j_ref_a_col: str = ""
    j_ref_b_col: str = ""

    @classmethod
    def from_dict(cls, d: dict) -> "JunctionConfig":
        cfg = cls(
            a_source_col=str(d.get("a_source_col", "")),
            b_source_col=str(d.get("b_source_col", "")),
            a_ext_col=str(d.get("a_ext_col", "")),
            b_ext_col=str(d.get("b_ext_col", "")),
            j_key_col=str(d.get("j_key_col", "")),
            key_template=str(d.get("key_template", "{A}-{B}") or "{A}-{B}"),
            b_regex_pattern=str(d.get("b_regex_pattern", "")),
            b_regex_replacement=str(d.get("b_regex_replacement", "")),
            required_col=str(d.get("required_col", "")),
            j_ref_a_col=str(d.get("j_ref_a_col", "")),
            j_ref_b_col=str(d.get("j_ref_b_col", "")),
        )
        for name, v in (("a_source_col", cfg.a_source_col),
                        ("b_source_col", cfg.b_source_col),
                        ("a_ext_col", cfg.a_ext_col),
                        ("b_ext_col", cfg.b_ext_col),
                        ("j_key_col", cfg.j_key_col)):
            if not v:
                raise DiffDeskError(f"多対多検証の設定 {name} が未指定です。")
        if "{A}" not in cfg.key_template or "{B}" not in cfg.key_template:
            raise DiffDeskError(
                "複合キーの形式には {A} と {B} の両方を含めてください(例: REL-{A}-{B})。")
        if cfg.b_regex_pattern:
            try:
                re.compile(cfg.b_regex_pattern)
            except re.error as e:
                raise DiffDeskError(f"親Bの正規化の正規表現が不正です: {e}")
        return cfg


def _macro(name: str, expected_keys: set[str], actual_keys: list[str],
           normalizer) -> dict:
    """期待キー集合と実績キー一覧の突合(件数+欠落/過剰の内訳)。"""
    actual_set: set[str] = set()
    dup_actual = 0
    for k in actual_keys:
        nk = normalizer(k)
        if not nk:
            continue
        if nk in actual_set:
            dup_actual += 1
        actual_set.add(nk)
    missing = sorted(expected_keys - actual_set)
    extra = sorted(actual_set - expected_keys)
    return {
        "name": name,
        "expected": len(expected_keys),
        "actual": len(actual_keys),
        "actual_unique": len(actual_set),
        "diff": len(actual_keys) - len(expected_keys),
        "missing": missing[:_SAMPLE_LIMIT],
        "missing_count": len(missing),
        "extra": extra[:_SAMPLE_LIMIT],
        "extra_count": len(extra),
        "dup_actual": dup_actual,
        "passed": not missing and not extra and dup_actual == 0,
    }


def verify_junction(source: Table, extract_a: Table, extract_b: Table,
                    extract_j: Table, config: JunctionConfig,
                    options: DiffOptions | None = None) -> dict:
    """多対多モデルの移行検証を実行し、4ステップの結果を返す。"""
    norm = make_normalizer(options or DiffOptions())

    ia = source.col_index(config.a_source_col)
    ib = source.col_index(config.b_source_col)
    ireq = source.col_index(config.required_col) if config.required_col else None
    iea = extract_a.col_index(config.a_ext_col)
    ieb = extract_b.col_index(config.b_ext_col)
    iej = extract_j.col_index(config.j_key_col)

    def norm_b(v: str) -> str:
        v = norm(v)
        if config.b_regex_pattern:
            v = re.sub(config.b_regex_pattern, config.b_regex_replacement, v)
        return v

    # 移行元から期待値を組み立てる
    a_expected: set[str] = set()
    b_expected: set[str] = set()
    j_expected: dict[str, tuple[str, str]] = {}  # 複合キー -> (親Aキー, 親Bキー)
    skipped_required = 0
    skipped_empty_pair = 0
    for row in source.rows:
        a_raw = norm(row[ia])
        b_raw = norm_b(row[ib])
        if a_raw:
            a_expected.add(a_raw)
        if b_raw:
            b_expected.add(b_raw)
        # 中間: スキップ条件(必須列が空白)と、キー成分が欠けた行は対象外
        if ireq is not None and not norm(row[ireq]):
            skipped_required += 1
            continue
        if not a_raw or not b_raw:
            skipped_empty_pair += 1
            continue
        jkey = config.key_template.replace("{A}", a_raw).replace("{B}", b_raw)
        j_expected.setdefault(jkey, (a_raw, b_raw))

    # ステップ1〜3: マクロ検証
    step1 = _macro("親A", a_expected, [r[iea] for r in extract_a.rows], norm)
    step2 = _macro("親B", b_expected, [r[ieb] for r in extract_b.rows], norm_b)
    step3 = _macro("中間", set(j_expected), [r[iej] for r in extract_j.rows], norm)
    step3["skipped_required"] = skipped_required
    step3["skipped_empty_pair"] = skipped_empty_pair

    # ステップ4: 未取込中間キーの原因切り分け
    a_actual = {norm(r[iea]) for r in extract_a.rows} - {""}
    b_actual = {norm_b(r[ieb]) for r in extract_b.rows} - {""}
    j_actual = {norm(r[iej]) for r in extract_j.rows} - {""}
    causes = {"missing_a": 0, "missing_b": 0, "missing_both": 0, "parents_ok": 0}
    samples: list[dict] = []
    for jkey, (a_key, b_key) in j_expected.items():
        if norm(jkey) in j_actual:
            continue
        a_ok = a_key in a_actual
        b_ok = b_key in b_actual
        cause = ("parents_ok" if a_ok and b_ok
                 else "missing_both" if not a_ok and not b_ok
                 else "missing_a" if not a_ok
                 else "missing_b")
        causes[cause] += 1
        if len(samples) < _SAMPLE_LIMIT:
            samples.append({"key": jkey, "a": a_key, "b": b_key, "cause": cause})

    total_missing = sum(causes.values())
    bottleneck = None
    if total_missing:
        worst = max(("missing_a", "missing_b", "missing_both", "parents_ok"),
                    key=lambda c: causes[c])
        bottleneck = worst if causes[worst] else None

    # 参照整合(任意): 中間抽出の親参照列が親抽出に実在するか
    ref_errors = None
    if config.j_ref_a_col and config.j_ref_b_col:
        ira = extract_j.col_index(config.j_ref_a_col)
        irb = extract_j.col_index(config.j_ref_b_col)
        bad_a = sum(1 for r in extract_j.rows if norm(r[ira]) not in a_actual)
        bad_b = sum(1 for r in extract_j.rows if norm_b(r[irb]) not in b_actual)
        ref_errors = {"bad_ref_a": bad_a, "bad_ref_b": bad_b}

    passed = step1["passed"] and step2["passed"] and step3["passed"]
    return {
        "parent_a": step1,
        "parent_b": step2,
        "junction": step3,
        "orphans": {"causes": causes, "total": total_missing,
                    "bottleneck": bottleneck, "samples": samples},
        "ref_errors": ref_errors,
        "passed": passed,
    }


CAUSE_JA = {
    "missing_a": "親Aの外部IDが移行先に存在しない",
    "missing_b": "親Bの外部IDが移行先に存在しない",
    "missing_both": "親A・親Bの両方が存在しない",
    "parents_ok": "両親は存在するが中間だけ未取込",
}


def build_orphan_table(result: dict, config: JunctionConfig) -> Table:
    """未取込の中間キー一覧をCSV出力用Tableにする(原因つき・アクション可能)。"""
    rows = [[s["key"], s["a"], s["b"], CAUSE_JA.get(s["cause"], s["cause"])]
            for s in result["orphans"]["samples"]]
    return Table(
        columns=["複合キー", "親Aキー", "親Bキー", "未取込の原因"],
        rows=rows, name="未取込一覧",
    )
